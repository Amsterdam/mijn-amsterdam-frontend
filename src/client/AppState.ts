import { createContext } from 'react';

import type {
  ServiceID,
  ServicesTips,
  ServicesType,
} from '../server/services/controller';
import { FeatureToggle } from '../universal/config';
import { apiPristineResult, ApiResponse } from '../universal/helpers/api';

export type AppState = {
  [key in ServiceID]: ApiResponse<
    ReturnTypeAsync<ServicesType[key]>['content']
  >;
} & {
  TIPS: ServicesTips | ApiResponse<any>;
};

export const PRISTINE_APPSTATE: AppState = {
  // Generated
  TIPS: apiPristineResult([]),
  NOTIFICATIONS: apiPristineResult([]),
  CASES: apiPristineResult([]),

  // Direct
  FOCUS_SPECIFICATIES: apiPristineResult({
    jaaropgaven: [],
    uitkeringsspecificaties: [],
  }),
  KREFIA: apiPristineResult(null, FeatureToggle.krefiaActive),
  FOCUS_AANVRAGEN: apiPristineResult([]),
  FOCUS_TOZO: apiPristineResult([]),
  FOCUS_TONK: apiPristineResult([]),
  FOCUS_BBZ: apiPristineResult([]),
  FOCUS_STADSPAS: apiPristineResult(null),
  WMO: apiPristineResult([]),
  ERFPACHT: apiPristineResult({ isKnown: false }),
  BELASTINGEN: apiPristineResult({ isKnown: true }),
  MILIEUZONE: apiPristineResult({ isKnown: false }),
  AKTES: apiPristineResult([], FeatureToggle.aktesActive),
  TOERISTISCHE_VERHUUR: apiPristineResult({
    vergunningen: [],
    registraties: [],
    daysLeft: 0,
  }),
  VERGUNNINGEN: apiPristineResult([]),

  // KVK / Handelsregister
  KVK: apiPristineResult(null),

  // Related
  BRP: apiPristineResult(null),
  AFVAL: apiPristineResult([]),
  AFVALPUNTEN: apiPristineResult({
    centers: [],
    datePublished: '',
  }),
  MY_LOCATION: apiPristineResult(null),

  // CMS content
  CMS_CONTENT: apiPristineResult({
    generalInfo: null,
    footer: null,
  }),
  CMS_MAINTENANCE_NOTIFICATIONS: apiPristineResult([]),
};

export const ALL_ERROR_STATE_KEY = 'ALL';

export function createAllErrorState(state: AppState, message: string) {
  return Object.assign({}, state, {
    [ALL_ERROR_STATE_KEY]: {
      status: 'ERROR',
      message,
    },
  });
}

export const AppContext = createContext<AppState>(PRISTINE_APPSTATE);
