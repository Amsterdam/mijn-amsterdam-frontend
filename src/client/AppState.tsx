import React, { createContext } from 'react';
import {
  ServicesDirectData,
  useServicesDirect,
} from './hooks/api/api.services-direct';
import { ServicesMapData, useServicesMap } from './hooks/api/api.services-map';
import {
  ServicesRelatedData,
  useServicesRelated,
} from './hooks/api/api.services-related';
import {
  ServicesTipsData,
  useServicesTips,
} from './hooks/api/api.services-tips';
import {
  ServicesUpdatesData,
  useServicesUpdates,
} from './hooks/api/api.services-updates';

import { ChaptersState } from './config/myChapters';
import { ComponentChildren } from '../universal/types/App.types';

export interface AppState extends ServicesRelatedData, ServicesDirectData {
  TIPS: ServicesTipsData;
  CHAPTERS: ChaptersState;
  BUURT: ServicesMapData;
  UPDATES: ServicesUpdatesData;
}
export type StateKey = keyof AppState;

export const AppContext = createContext<AppState>({} as AppState);

interface AppStateProps {
  children: ComponentChildren;
}

interface MockAppStateProps {
  children: ComponentChildren;
  value: Partial<AppState>;
}

export function useAppState(): AppState {
  const servicesRelatedApi = useServicesRelated();
  const servicesDirectApi = useServicesDirect();
  const tipsApi = useServicesTips();
  const updatesApi = useServicesUpdates();
  const mapData = useServicesMap();

  return {
    ...servicesRelatedApi.data,
    ...servicesDirectApi.data,
    TIPS: tipsApi.data,
    UPDATES: updatesApi.data,
    CHAPTERS: {
      isLoading: true,
      items: [],
    },
    BUURT: mapData,
  };
}

export function MockAppStateProvider({ children, value }: MockAppStateProps) {
  return (
    <AppContext.Provider value={value as AppState}>
      {children}
    </AppContext.Provider>
  );
}

export default function AppStateProvider({ children }: AppStateProps) {
  const appState = useAppState();
  return <AppContext.Provider value={appState}>{children}</AppContext.Provider>;
}
