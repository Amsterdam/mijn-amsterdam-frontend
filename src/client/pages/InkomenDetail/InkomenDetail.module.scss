@import '../../styles/_global.scss';

.DetailPageContent {
  min-height: 4rem;
  padding-bottom: 5rem;
}

.Tozo2Alert {
  p {
    margin-top: 0;
  }
}
