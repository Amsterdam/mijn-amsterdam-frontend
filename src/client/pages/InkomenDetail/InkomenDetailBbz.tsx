import { useCallback } from 'react';
import { useParams } from 'react-router-dom';
import Linkd from '../../components/Button/Button';
import { ExternalUrls } from '../../config/app';
import { useAppStateGetter } from '../../hooks';
import StatusDetail, { StatusSourceItem } from '../StatusDetail/StatusDetail';

export default function InkomenDetailBbz() {
  const { WPI_BBZ } = useAppStateGetter();
  const { id } = useParams<{ id: string }>();
  const statusItem = WPI_BBZ.content?.find((item) => item.id === id);
  const hasDecisionStep =
    statusItem?.steps.some((step) => step.id.includes('besluit')) ?? false;

  const pageContent = useCallback(
    (isLoading: boolean, inkomenItem: StatusSourceItem) => {
      return (
        <>
          <p>
            Hieronder ziet u de status van uw aanvraag voor een uitkering of
            lening van het Bbz. Ook als u een IOAZ uitkering heeft aangevraagd
            ziet u de status hieronder. Als u meerdere aanvragen voor het Bbz
            hebt gedaan, dan krijgt u 1 besluit als antwoord op al uw aanvragen
            voor het Bbz. Het duurt maximaal 3 werkdagen voordat uw documenten
            over het Bbz in Mijn Amsterdam staan.
          </p>
          <p>
            Hebt u schuldhulp aangevraagd? Dan wordt daarover contact met u
            opgenomen.
          </p>
          <p>
            <Linkd external={true} href={ExternalUrls.WPI_BBZ}>
              Meer informatie over het Bbz
            </Linkd>
          </p>
        </>
      );
    },
    []
  );

  return (
    <StatusDetail
      thema="INKOMEN"
      stateKey="WPI_BBZ"
      pageContent={pageContent}
      maxStepCount={() => -1}
      highlightKey={false}
      statusLabel={() => `Bbz-aanvraag`}
      showStatusLineConnection={!hasDecisionStep} // There is no logical connection between the historic BBZ steps, therefor we do not show the checkmarks as they imply a linear proces.
      documentPathForTracking={(document) =>
        `/downloads/inkomen/bbz/${document.title.split(/\n/)[0]}`
      }
      reverseSteps={hasDecisionStep} // For an unknown business reason, the historic steps of BBZ are shown in reverse.
    />
  );
}
