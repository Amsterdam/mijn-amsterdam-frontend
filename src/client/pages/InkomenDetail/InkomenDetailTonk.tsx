import { useCallback } from 'react';
import Linkd from '../../components/Button/Button';
import { ExternalUrls } from '../../config/app';
import StatusDetail from '../StatusDetail/StatusDetail';

export default function InkomenDetailTonk() {
  const pageContent = useCallback((isLoading, inkomenItem) => {
    return (
      <>
        <p>
          Hieronder ziet u de status van uw aanvraag TONK. Als u meerdere
          aanvragen voor de TONK hebt gedaan, dan krijgt u 1 besluit als
          antwoord op al deze aanvragen. Het duurt maximaal 3 werkdagen voordat
          uw documenten over de TONK in Mijn Amsterdam staan.
        </p>
        <p>
          <Linkd external={true} href={ExternalUrls.WPI_TONK}>
            Meer informatie over de TONK
          </Linkd>
        </p>
      </>
    );
  }, []);

  return (
    <StatusDetail
      chapter="INKOMEN"
      stateKey="WPI_TONK"
      showToggleMore={false}
      pageContent={pageContent}
      maxStepCount={() => -1}
      highlightKey={false}
      statusLabel={(statusItem) => `${statusItem?.about}-aanvraag`}
      documentPathForTracking={(document) =>
        `/downloads/inkomen/tonk/${document.title.split(/\n/)[0]}`
      }
    />
  );
}
