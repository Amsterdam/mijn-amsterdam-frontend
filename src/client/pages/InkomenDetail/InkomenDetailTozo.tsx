import { useCallback } from 'react';
import Linkd from '../../components/Button/Button';
import { ExternalUrls } from '../../config/app';
import StatusDetail, { StatusSourceItem } from '../StatusDetail/StatusDetail';
import {
  Grid,
  Link,
  LinkList,
  Paragraph,
} from '@amsterdam/design-system-react';

export default function InkomenDetailTozo() {
  const pageContent = useCallback(
    (isLoading: boolean, inkomenItem: StatusSourceItem) => {
      return (
        <>
          <Grid.Cell span="all">
            <Paragraph>
              Hieronder ziet u de status van uw aanvraag voor de{' '}
              {inkomenItem?.about || 'Tozo'}. Als u meerdere aanvragen voor de{' '}
              {inkomenItem?.about || 'Tozo'} hebt gedaan, dan krijgt u 1 besluit
              als antwoord op al uw aanvragen voor de{' '}
              {inkomenItem?.about || 'Tozo'}. Het duurt maximaal 3 werkdagen
              voordat uw documenten over de {inkomenItem?.about || 'Tozo'} in
              Mijn Amsterdam staan.
            </Paragraph>
          </Grid.Cell>
          <Grid.Cell span="all">
            <LinkList>
              <LinkList.Link rel="noreferrer" href={ExternalUrls.WPI_TOZO}>
                Meer informatie over de Tozo
              </LinkList.Link>
            </LinkList>
          </Grid.Cell>
        </>
      );
    },
    []
  );

  return (
    <StatusDetail
      thema="INKOMEN"
      stateKey="WPI_TOZO"
      pageContent={pageContent}
      maxStepCount={() => -1}
      highlightKey={false}
      statusLabel={(statusItem) => `${statusItem?.about}-aanvraag`}
      documentPathForTracking={(document) =>
        `/downloads/inkomen/tozo/${document.title.split(/\n/)[0]}`
      }
    />
  );
}
