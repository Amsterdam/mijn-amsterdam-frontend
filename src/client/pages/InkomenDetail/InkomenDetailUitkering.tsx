import { useCallback } from 'react';
import Linkd from '../../components/Button/Button';
import { ExternalUrls } from '../../config/app';
import StatusDetail, { StatusSourceItem } from '../StatusDetail/StatusDetail';

export const MAX_STEP_COUNT_WPI_REQUEST = 4;

export default function InkomenDetailUitkering() {
  const pageContent = useCallback(
    (isLoading: boolean, inkomenItem: StatusSourceItem) => {
      return (
        <>
          <p>
            Hieronder ziet u de status van uw aanvraag voor een
            bijstandsuitkering. Het duurt maximaal 3 werkdagen voordat uw
            documenten over de bijstandsuitkering in Mijn Amsterdam staan.
          </p>
          <p>
            <Linkd external={true} href={ExternalUrls.WPI_BIJSTANDSUITKERING}>
              Meer informatie over de bijstandsuitkering
            </Linkd>
          </p>
        </>
      );
    },
    []
  );

  return (
    <StatusDetail
      chapter="INKOMEN"
      stateKey="WPI_AANVRAGEN"
      pageContent={pageContent}
      maxStepCount={(hasDecision) =>
        !hasDecision ? MAX_STEP_COUNT_WPI_REQUEST : undefined
      }
      documentPathForTracking={(document) =>
        `/downloads/inkomen/bijstandsuitkering/${document.title.replace(
          /\\n/,
          ''
        )}`
      }
    />
  );
}
