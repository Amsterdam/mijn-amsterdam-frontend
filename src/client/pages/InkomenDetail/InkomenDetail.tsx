import {
  Alert,
  ChapterIcon,
  DetailPage,
  LoadingContent,
  PageContent,
  PageHeading,
  StatusLine,
} from '../../components';
import { AppRoutes, ChapterTitles } from '../../../universal/config';
import React, { useContext } from 'react';
import {
  StatusLineItem,
  StepType,
} from '../../components/StatusLine/StatusLine';
import { isError, isLoading } from '../../../universal/helpers';

import { AppContext } from '../../AppState';
import styles from './InkomenDetail.module.scss';
import useRouter from 'use-react-router';

export function altDocumentContent(
  statusLineItem: StatusLineItem,
  stepNumber: number
) {
  if (!!statusLineItem.documents.length) {
    return '';
  }

  if (
    statusLineItem.status === 'Meer informatie nodig' &&
    statusLineItem.isRecent &&
    !statusLineItem.isLastActive
  ) {
    return <b>U heeft deze brief per post ontvangen.</b>;
  }

  return ['Meer informatie nodig', 'Besluit'].includes(
    statusLineItem.status
  ) ? (
    statusLineItem.isRecent ? (
      <b>
        U ontvangt $
        {statusLineItem.status === 'Besluit' ? 'dit besluit' : 'deze brief'} per
        post.
      </b>
    ) : (
      <b>
        U heeft $
        {statusLineItem.status === 'Besluit' ? 'dit besluit' : 'deze brief'} per
        post ontvangen.
      </b>
    )
  ) : (
    ''
  );
}

export default () => {
  const { FOCUS } = useContext(AppContext);

  const {
    match: {
      params: { id },
    },
  } = useRouter();

  const FocusItem = FOCUS?.aanvragen.find(item => item.id === id);
  const noContent = !isLoading(FOCUS?.aanvragen) && !FocusItem;
  const lineItemsTotal = FocusItem?.process.length || 0;
  const items =
    FocusItem?.process.map((item, index) => {
      const stepType: StepType =
        index === lineItemsTotal
          ? 'last-step'
          : index === 0
          ? 'first-step'
          : 'intermediate-step';
      return Object.assign(item, { stepType });
    }) || [];

  return (
    <DetailPage>
      <PageHeading
        icon={<ChapterIcon />}
        backLink={{ to: AppRoutes.INKOMEN, title: ChapterTitles.INKOMEN }}
      >
        {FocusItem && FocusItem.title}
      </PageHeading>
      <PageContent className={styles.DetailPageContent}>
        {(isError(FOCUS?.aanvragen) || noContent) && (
          <Alert type="warning">
            <p>We kunnen op dit moment geen gegevens tonen.</p>
          </Alert>
        )}
        {isLoading && <LoadingContent />}
      </PageContent>
      {!!FocusItem && (
        <StatusLine
          trackCategory={`Inkomen en Stadspas / ${FocusItem.productTitle}`}
          items={items}
          altDocumentContent={altDocumentContent}
          id={id}
        />
      )}
    </DetailPage>
  );
};
