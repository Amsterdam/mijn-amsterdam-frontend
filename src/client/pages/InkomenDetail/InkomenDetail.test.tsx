import { shallow } from 'enzyme';
import React from 'react';
import { AppState } from '../../AppState';
import { MockAppStateProvider } from '../../AppStateProvider';
import InkomenDetail from './InkomenDetail';

const APP_STATE = { WMO: { content: [], status: 'OK' } } as Partial<AppState>;

it('Renders without crashing', () => {
  shallow(
    <MockAppStateProvider value={APP_STATE}>
      <InkomenDetail />
    </MockAppStateProvider>
  );
});
