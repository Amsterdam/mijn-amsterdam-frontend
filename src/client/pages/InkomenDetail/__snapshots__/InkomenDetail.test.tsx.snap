// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InkomenDetail /> Matches the Full Page snapshot 1`] = `
<DocumentFragment>
  <main
    class="Page DetailPage"
  >
    <header
      class="PageHeading hasIcon hasBackLink"
    >
      <span
        aria-hidden="true"
        class="Icon"
      >
        <svg
          aria-label="Inkomen en Stadspas"
          class="ChapterIcon"
          fill="black"
        >
          inkomen.svg
        </svg>
      </span>
      <div
        class="HeadingInner"
      >
        <a
          class="Button Button__plain Button__lean Linkd BackLink"
          href="/inkomen-en-stadspas"
        >
          <svg
            aria-hidden="true"
            class="Icon Icon__left"
            fill="#000"
            height="14px"
            width="14px"
          >
            Chevron-Left.svg
          </svg>
          Inkomen en Stadspas
        </a>
        <h2
          class="Heading Heading__large"
        >
          Aanvraag inkomen item
        </h2>
      </div>
    </header>
    <div
      class="PageContent DetailPageContent"
    />
    <div
      class="StatusLine"
    >
      <h4
        class="Heading Heading__tiny ListHeading"
      >
        Status
      </h4>
      <ul
        class="List"
      >
        <li
          class="ListItem ListItem--highlight"
          style="display: block;"
        >
          <div
            class="ListItemInner"
          >
            <div
              aria-label="Vinkje, processtap gereed"
              class="StatusConnection"
            >
              <span
                class="ConnectLine ConnectLine--start"
              />
              <span
                class="Checkmark Checkmark--active Checkmark--checked"
              />
            </div>
            <div
              class="Panel Panel--status"
            >
              <strong
                class="StatusTitle"
              >
                aanvraag
              </strong>
              <time
                class="StatusDate"
              >
                24 juli 2020
              </time>
            </div>
            <div
              class="Panel Panel--description"
            >
              <div
                class="PanelContent"
              >
                <p>
                  Uw aanvraag is binngekomen
                </p>
              </div>
            </div>
            <div
              class="Panel Panel--documents"
            >
              <ul
                class="DocumentList"
              >
                <li
                  class="DocumentListItem"
                >
                  <span
                    class="DocumentLinkWrap"
                  >
                    <a
                      class="Button Button__plain Button__lean Linkd DocumentLink"
                      href="/inkomen-en-stadspas/bijstandsuitkering/http://example.org/blaat"
                    >
                      <svg
                        aria-hidden="true"
                        class="Icon Icon__left"
                        fill="#000"
                        height="14px"
                        width="14px"
                      >
                        Download.svg
                      </svg>
                      Documentje
                    </a>
                  </span>
                </li>
              </ul>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </main>
</DocumentFragment>
`;
